webapp/sources/rudder/rudder-core/src/main/scala/com/normation/cfclerk/services/impl/GitTechniqueReader.scala

```diff
@@ -488,80 +535,115 @@ class GitTechniqueReader(
 
   }
 
-  private[this] def processTechniques(revTreeId: ObjectId, techniqueInfos : InternalTechniquesInfo, parseDescriptor:Boolean) : Unit = {
+  private[this] def processTechniques(gitRepo: Repository, revTreeId: ObjectId, techniqueInfosRef: Ref[InternalTechniquesInfo], parseDescriptor: Boolean) : IOResult[Unit] = {
       //a first walk to find categories
-      val tw = new TreeWalk(repo.db.runNow)
+    val managed = ZManaged.make(IOResult.effect {
+      val tw = new TreeWalk(gitRepo)
       tw.setFilter(new FileTreeFilter(canonizedRelativePath, techniqueDescriptorName))
       tw.setRecursive(true)
       tw.reset(revTreeId)
-
-      //now, for each potential path, look if the cat or policy
-      //is valid
-      while(tw.next) {
-        val path = toTechniquePath(tw.getPathString) //we will need it to build the category id
-        processTechnique(repo.db.runNow.open(tw.getObjectId(0)).openStream, path.path, techniqueInfos, parseDescriptor, revTreeId).either.runNow match {
-          case Left(err)  => logger.error(s"Error with technique at path: '${path.path}', it will be ignored. Error: ${err.fullMsg}")
-          case Right(()) => // ok
-        }
+      tw
```

Comment:
In place of `while` loop, we use a recursive function guarded by the while condition. We just need to be carefull to use `IOResult.effect` everywhere because most things throw exception, and to not use anything `tw` related before the first call to `next` (because in java-logic, it's that call which init things)